ops/lib/__init__.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+import re
+
+from pkgutil import get_importer
+from importlib.util import module_from_spec
+from functools import total_ordering
+
+
+_libraries = {}
+
+_libline_re = re.compile(r'''^LIB([A-Z]+)\s+=\s+([0-9]+|['"][a-zA-Z0-9_.-@]+['"])\s*$''')
+_libname_re = re.compile(r'''^[a-z][a-z0-9]+$''')
+
+# Not perfect, but should do for now.
+_libauthor_re = re.compile(r'''^[A-Za-z0-9_+.-]+@[a-z0-9_-]+(?:\.[a-z0-9_-]+)*\.[a-z]{2,3}$''')
+
+
+def use(name, author, api):
+    if not _libname_re.match(name):
+        raise ValueError("invalid library name: {} (chars and digits only)".format(name))
+    if not _libauthor_re.match(author):
+        raise ValueError("invalid library author email: {}".format(author))
+    if not isinstance(api, int):
+        raise ValueError("invalid library API: {} (must be an int)".format(api))
+
+    versions = _libraries.get((name, author), ())
+    for lib in versions:
+        if lib.api == api:
+            return lib.import_module()
+
+    others = ', '.join(str(lib.api) for lib in versions)
+    if others:
+        raise RuntimeError('cannot find "{}" from {} with API {} (have {})'.format(
+            name, author, api, others))
+    else:
+        raise RuntimeError('cannot find library "{}" from {}'.format(
+            name, author))
+
+
+def autoimport():
+    _libraries.clear()
+    for sys_dir in sys.path:
+        if not isinstance(sys_dir, str):
```

Comment:
The are always strings. In the very unprobable event that somebody put non-strings inside, it will be caught in the `except OSError` below, right?